Use auto-deploying pipeline for staging and production

alexiswl opened the pull request:
# ccc3fe28f023927af444626bab65e249f7068fd9

> Update stack to use auto-deploying pipeline for staging and production

## Updated files

### bin/nextflow-stack.ts

* Updated app list to have dev docker build and dev application stacks (this way the two can be deployed separately if needed)

* Added in the build-pipeline-stack for staging and production.

### constants.ts

* Added in environment objects
* Added in ssm parameters for dev, stg and prod

### lib / oncoanalyser-stack.ts

* Highlighted hardcoded bucket names and prefixes - to be fixed later on
* Added in role name as a stack output (used to set an ssm parameter in the parent application stack)

### package.json

* Upgraded to latest version of cdk

## New Files

### lib/application-stack.ts

* Wraps the Shared stack and OncoAnalyser stacks together and also builds the ssm parameters needed for the run script

* Build Stage wraps around the application stack as a codepipeline stack

### lib/docker-build-stack

* An ecr deployment stack that takes a path to a DockerFile is stored and uploads to the appropriate repository.

* For dev this the dev account, for staging and prod this is the bastion account.  We will need to allow access to this repository from staging and production

### lib/pipeline-stack.ts

* Self-mutating pipeline deployed into the bastion account. Builds the docker image in bastion and then deploys the stacks into staging and prod (with a manual push-to-prod setting in between the two deployments)

* Uses some hacky trick to collect the codestar commit id

### package-lock.json

Standard package lock file

# 3720786cdb72f55fe0e641d8ffdf9d763702619d

> Remove hardcoded paths in run.sh and nextflow config

## Updated files

### Nextflow.config

Move nextflow.config to a template instead then 'sadly' use some ugly (as pretty as we can make them) sed hacks to update the config in a temp file. 

### Run.sh

Use ssm parameters instead to collect required bucket and instance profile names

General clean up of run.sh as well to prevent shellcheck from complaining

### DockerFile

* General tidyup
* Set software and dep versions as ARGs instead

alexiswl committed:
Update stack to use auto-deploying pipeline for staging and produciton

Also made additional file changes

## Updated files

### bin/nextflow-stack.ts

* Updated app list to have dev docker build and dev application stacks (this way the two can be deployed separately if needed)

* Added in the build-pipeline-stack for staging and production.

### constants.ts

* Added in environment objects
* Added in ssm parameters for dev, stg and prod

### lib / oncoanalyser-stack.ts

* Highlighted hardcoded bucket names and prefixes - to be fixed later on
* Added in role name as a stack output (used to set an ssm parameter in the parent application stack)

### package.json

* Upgraded to latest version of cdk

## New Files

### lib/application-stack.ts

* Wraps the Shared stack and OncoAnalyser stacks together and also builds the ssm parameters needed for the run script

* Build Stage wraps around the application stack as a codepipeline stack

### lib/docker-build-stack

* An ecr deployment stack that takes a path to a DockerFile is stored and uploads to the appropriate repository.

* For dev this the dev account, for staging and prod this is the bastion account.  We will need to allow access to this repository from staging and production

### lib/pipeline-stack.ts

* Self-mutating pipeline deployed into the bastion account. Builds the docker image in bastion and then deploys the stacks into staging and prod (with a manual push-to-prod setting in between the two deployments)

* Uses some hacky trick to collect the codestar commit id

### package-lock.json

Standard package lock file

alexiswl committed:
Update run.sh and config to cater for multi-account deployments

Remove hardcoded paths in run.sh and nextflow config

## Nextflow.config

Move nextflow.config to a template instead then 'sadly' use some ugly (as pretty as we can make them) sed hacks to update the config in a temp file. 

## Run.sh

Use ssm parameters instead to collect required bucket and instance profile names

General clean up of run.sh as well to prevent shellcheck from complaining

## DockerFile

* General tidyup
* Set software and dep versions as ARGs instead

scwatts commented on docker-images/oncoanalyser/Dockerfile:
```diff
@@ -1,56 +1,75 @@
-RUN \
-  wget -q -P /usr/bin/ 'https://stratus-documentation-us-east-1-public.s3.amazonaws.com/cli/1.2.0/linux/ica' && \
-  chmod 755 /usr/bin/ica
-COPY assets/ica_config.yaml /root/.ica/config.yaml
```
Should also be safe to delete `assets/ica_config.yaml`.

alexiswl replied:
Resolved in f69c5dce51b421887ed895a06479ffe82c2428a3

scwatts commented on docker-images/oncoanalyser/assets/nextflow_aws_template.config:
```diff
@@ -12,21 +12,21 @@ wave {
+      bwa_index       = "__S3_HMF_GENOMES_DATA_PATH__/GRCh38_umccr/bwa_index/0.7.17-r1188/"
+      bwa_index_image = "__S3_HMF_GENOMES_DATA_PATH__/GRCh38_umccr/bwa_index_image/0.7.17-r1188/GRCh38_full_analysis_set_plus_decoy_hla.fa.img"
+      gridss_index    = "__S3_HMF_GENOMES_DATA_PATH__/GRCh38_umccr/gridss_index/2.13.2/GRCh38_full_analysis_set_plus_decoy_hla.fa.gridsscache"
+      star_index      = "__S3_HMF_GENOMES_DATA_PATH__/GRCh38_umccr/star_index/gencode_38/2.7.3a/"
```
Since these reference files aren't Hartwig specific can `__S3_HMF_GENOMES_DATA_PATH__` be changed to `__S3_GENOMES_DATA_PATH__`?

alexiswl replied:
Resolved in 7acc9c13da2d7887fd0065cbb5f7971d1f22d4dc

scwatts commented on docker-images/oncoanalyser/assets/run.sh:
```diff
@@ -165,79 +177,175 @@ fi
+
+get_hmf_genomes_path_from_ssm(){
+  get_ssm_parameter_value "/oncoanalyser/reference_data/genomes"
+}
```
Also dropping `hmf` for both the VIRUSBreakend database and reference genomes function name would be helpful here.

alexiswl replied:
Resolved in 7acc9c13da2d7887fd0065cbb5f7971d1f22d4dc

scwatts commented on docker-images/oncoanalyser/assets/run.sh:
```diff
@@ -165,79 +177,175 @@ fi
-      --function-name arn:aws:lambda:ap-southeast-2:472057503814:function:IcaSecretsPortalProvider \
+    AWS_REGION=ap-southeast-2 aws lambda invoke \
+      --function-name "$(get_ica_secrets_portal_provider_arn)" \
       response.json 1>/dev/null
```
The IcaSecretsPortalProvider Lambda only exists in prod and has been used to grant read permissions on the GDS production volume for oncoanalyser running dev. Having the CodePipeline that allows us to properly deploy into stg/prod means that we no longer need that work around and instead can use the IcaSecretsPortal token from Secrets Manager.

Leaving it open to you whether you'd like to make that change, otherwise I'll update outside of this PR.

alexiswl replied:
Updated in 7acc9c13da2d7887fd0065cbb5f7971d1f22d4dc to use the secret rather than the lambda function to collect the access token

scwatts commented on docker-images/oncoanalyser/assets/run.sh:
```diff
@@ -248,17 +356,88 @@ region = ap-southeast-2
+# using Fusion, S3 output directory, etc), we must set the EC2 instance IAM role to a profile with such permissions.
+# Here I associate the instance with the OncoanalyserStack task role. There may be better approaches to achieve this.
+
+# NOTE(SW): this is to be updated manually for now until build process is placed into CodePipeline
```
Can remove this comment now 😃

alexiswl replied:
Done in 7acc9c13da2d7887fd0065cbb5f7971d1f22d4dc

scwatts commented on docker-images/oncoanalyser/assets/run.sh:
```diff
@@ -346,31 +520,45 @@ elif [[ ${mode} == 'wgts_existing_both' ]]; then
+## CREATE NEXTFLOW CONFIG ##
+nextflow_config_tmp="$( \
+  mktemp --suffix "nextflow.config"\
+)"
```
I had been meaning to capture the Nextflow config file when uploading .nextflow/, .nextflow.log, and samplesheet.csv to give us better reproducibility for runs. Though I'm undecided whether this is excessive.

At least for now, can we use a stable filename (e.g. `nextflow.config`) and place it into the cwd?

alexiswl replied:
Updated in 7acc9c13da2d7887fd0065cbb5f7971d1f22d4dc to use filename _nextflow.config_ in current working directory.

scwatts reviewed (commented):
Amazing, thank you! I have only couple of very minor additions in the individual comments.

No problem merging with the FIXME comments since these are mostly related to account-specific paths and will be discussed + decided at a later stage

alexiswl committed:
Runtime changes


## Nextflow Template Config

Renamed S3_HMF_GENOMES_DATA_PATH placeholder by removing HMF midfix

## Run.sh

* Renamed get reference path functions by removing hmf midfix. 
* Use secrets manager to collect ICA_ACCESS_TOKEN
* Declare local variables for stage_gds_fp function and quote all vars

* Collect AWS REGION from metadata rather than have it hard-coded

* Save nextflow.config rather than have it stored as a temp file (good for reproducibility).

alexiswl committed:
Removed ica config yaml from docker assets

alexiswl committed:
Updated EOF lines

Some files didn't have a newline char at the end of the file. Updating for consistency

scwatts commented:
Missed this but can these also be updated?

https://github.com/umccr/nextflow-stack/blob/cd78f467ed90ab83cae83fa8516d3d86cc1f413c/constants.ts#L52-L58

alexiswl committed:
Remove HMF prefixes from virus breakend and genomes constants vars

alexiswl committed:
Resolved hardcoded bucket issue by using bucket name and prefix as inputs

alexiswl committed:
Rename constants

HMF_REFERENCE_DATA_PATH_DEV -> HMF_PATH_DEV, HMF_REFDATA_PATH_DEV
VIRUS_BREAKEND_DB_PATH_DEV -> VIRUSBREAKEND_DB_PATH_DEV
GENOMES_DATA_PATH_DEV -> GENOMES_PATH_DEV

alexiswl committed:
Added context json to gitignore

alexiswl committed:
Upgraded to aws cdk v2.73.0

alexiswl committed:
Renamed constant REFERENCE_DATA_PATH

Should have been HMF_REFDATA_PATH

alexiswl committed:
Use ssm parameter for docker tag / import restructure / secretmanger fix

* Removed unnecessary imports, import functions as needed
* Add secrets manager
* Fixed job definition access policy (only access job defs starting with nf-)
